mltsp/datasets/util.py

```diff
@@ -0,0 +1,74 @@
+import hashlib
+import os
+import tarfile
+
+import pandas as pd
+
+from .. import util
+
+try:
+    import urllib.request as request
+except:
+    import urllib2 as request
+
+
+def _md5sum_file(path):
+    """Calculate the MD5 sum of a file."""
+    with open(path, 'rb') as f:
+        m = hashlib.md5()
+        while True:
+            data = f.read(8192)
+            if not data:
+                break
+            m.update(data)
+    return m.hexdigest()
+
+
+def download_file(data_dir, base_url, filename):
+    """Download a single file into the given directory."""
+    file_path = os.path.join(data_dir, filename)
+    opener = request.urlopen(base_url + filename)
+    with open(file_path, 'wb') as f:
+        f.write(opener.read())
+    return file_path
+
+
+def download_and_extract_archives(data_dir, base_url, filenames, md5sums=None,
+                                 remove_archive=True):
+    """Download list of data archives, verify md5 checksums (if applicable),
```

Comment:
You could still use this for an arbitrary archive that you haven't seen before; also, for testing we send back fake data that we wouldn't want to have to keep re-generating MD5's for. But yeah, in practice you would always want to specify the hash.